### PR TITLE
Automatically create a link when selected text is a URL

### LINKED ISSUE #3513: Insert link: adding a link should auto-detect selected URLs
### Scenario

- Paste or type a URL into a Paragraph block
- Click the `Link` button in the toolbar

### Expected behavior

The link dialog detects that I have selected a URL in my post content and pre-fills the URL field with the selected text.  (Or, better yet, the URL is automatically recognized and turned into a link.)

### Actual behavior

The URL field in the link dialog is left blank.  If I have used my clipboard for anything else since then, I have to perform several extra operations to fill in the URL.

### PR DESCRIPTION
**Pull request by noisysocks**
Closes #3513. 

When a URL is selected and the Link button is clicked or the Link keyboard shortcut is pressed, a link with that URL set as its `href` should be instantly created.

auto-link

As part of this, I pulled an existing link regular expression out into its own `isURL` function that lives in the `@wordpress/url` package. Not sure if it's actually a useful helper function, though—happy to go back on this decision.

**To test:**

```
npm run test-e2e
```

Or, for those of us who distrust automation:

1. Highlight a URL
2. Click _Link_ or press <kbd>Cmd</kbd>+<kbd>K</kbd>
3. The selected URL should immediately become a link

### CODE REVIEW

File: packages/url/src/index.js [hunk 1]
```diff
@@ -3,9 +3,21 @@
...
 const USABLE_HREF_REGEXP = /^(?:[a-z]+:|#|\?|\.|\/)/i;
 
+/**
+ * Determines whether the given string looks like a URL.
+ *
+ * @param {string} url The string to scrutinise.
+ *
+ * @return {boolean} Whether or not it looks like a URL.
+ */
+export function isURL( url ) {
```
Comment (jorgefilipecosta) [inline packages/url/src/index.js]: Nice util, I think we may be able to use it in:
https://github.com/WordPress/gutenberg/blob/f9da76a18f1dda3b0aaa00218b49914901c6ab0e/packages/editor/src/components/rich-text/index.js#L285
Removing a duplicate regex.

**Review by jorgefilipecosta — COMMENTED** [review]
Nice feature 👍 
It is working well I think we have a small bug that we should address before the merge. After we add a link using this method if we click right away on the edit button nothing happens the edit button does not work. If we select other block and then come back and press edit things work well.

File: packages/url/src/test/index.test.js [hunk 2]
```diff
@@ -1,7 +1,42 @@
...
+		const urls = [
+			'http://wordpress.org',
+			'https://wordpress.org',
+			'HTTPS://WORDPRESS.ORG',
+			'https://wordpress.org/foo#bar',
+			'https://localhost/foo#bar',
+		];
+
+		forEach( urls, ( url ) => {
+			expect( isURL( url ) ).toBe( true );
```
[the remark this answers is not in the document]
Comment (noisysocks) [reply packages/url/src/test/index.test.js]: 👍 0d218ad96

File: packages/url/src/index.js (hunk 1 above)
Comment (noisysocks) [reply packages/url/src/index.js]: I've already replaced that regex 🙂

**Comment by noisysocks** [comment]
Good catch @jorgefilipecosta! I've fixed this in 50c9c01b0.

**Review by jorgefilipecosta — APPROVED** [review]
It is working well in my tests 👍

**Comment by mtias** [comment]
Thanks for this improvement!

**Pull request merged**